doc/man/st-info.1
.\" Automatically generated by Pandoc 2.4
.\"
.TH "st\-flash" "1" "Feb 2018" "Open Source STMicroelectronics Stlink Tools" "stlink"
.hy
.SH NAME
.PP
st\-info \- Provides information about connected STLink and STM32
devices
.SH SYNOPSIS
.PP
\f[I]st\-info\f[R] [\f[I]OPTIONS\f[R]]
.SH DESCRIPTION
.PP
Provides information about connected STLink programmers and STM32
devices: Serial code, openocd, flash, sram, page size, chipid,
description.
.SH OPTIONS
.TP
.B \[en]version
Print version information
.TP
.B \-\-flash
Display amount of flash memory available in the device
.TP
.B \-\-sram
Display amount of sram memory available in device
.TP
.B \-\-descr
Display textual description of the device
.TP
.B \-\-pagesize
Display the page size of the device
.TP
.B \-\-chipid
Display the chip ID of the device
.TP
.B \-\-serial
Display the serial code of the device
.TP
.B \-\-probe
Display the summarized information of the connected programmers and
devices
.SH EXAMPLES
.PP
Display information about connected programmers and devices
.IP
.nf
\f[C]
$ st\-info \-\-probe
\f[R]
.fi
.SH SEE ALSO
.PP
st\-util(1), st\-flash(1)
.SH COPYRIGHT
.PP
This work is copyrighted.
Stlink contributors.
See \f[I]LICENSE\f[R] file in the stlink source distribution.


ChangeLog
Stlink ChangeLog
================

v1.2.0
======

Release date: 16 may 2016

Features added:

* Add multiple stlink probing (`st-info --probe`, `st-info --hla-serial`) with printing serial in hex and OpenOCD `hla_serial` format (Jerry Jacobs)
* Add stlink usb probe API functions (Jerry Jacobs)
* Added parameter to specify one stlink v2 of many (Georg von Zengen)

Changes:

* Refactoring/fixes of flash loader (Maxime Coquelin)

Updates and fixes:

* Synchronize cache for stm32f7 (Tristan Gingold)
* Allow flashing of STM32L4 down to 1.71 V (Greg Meiste)
* Fix on stm32l4 to clear flash mass erase flags on CR (Bruno Dal Bo)
* Proper writing of page 0 of second bank for stm32l476xe (Tobias Badertscher)
* Trace the read data in stlink_read_debug32 and not the address of the variable (Tobias Badertscher)
* Mac OS X El Capitan platform support confirmation (Nikolay)
* Do not send a NUL at end of packets to gdb (Tristan Gingold)
* Correctly compute flash write size for partial pages (Dave Vandervies)
* _stlink_usb_reset use hardreset (mlundinse)
* Make sure MCU is halted before running RAM based flashloaders (mlundinse)
* Could not flash STM32_F3_SMALL (Max Chen)
* STM32F4 8-bit support for 1.8v operation (Andy Isaacson)
* Fix F2 memory map (Nicolas Schodet)
* Memory map for stm32f42xxx and stm32f43xxx devices (Craig Lilley)
* Stm32l0x flash loader (Robin Kreis)

Chip support added for:

* STM32L053R8 (Jean-Luc Béchennec)
* STM32F7 Support (mlundinse)
* Add STM32L4 to CHIPID #defines and devices[], flash driver and loaded (Dave Vandervies)
* Basic support for F446 (Pavel Kirienko)
* STM32F303 High Density
* STM32L1xx Cat.2 devices (Nicolas Schodet)

Board support added for:

* Nucleo-F303RE (Kyle Manna)
* Nucleo-F411RE (texane)

Build system:

* Travis: Initial support for Travis continues integration on Linux & Mac OS X (Jerry Jacobs)
* CMake: Document in README.md and add extra strict compiler flags (Jerry Jacobs)
* CMake: First stab at a cmake build (Josh Bialkowski)


src/st-info.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

#include <stlink-common.h>

static void usage(void)
{
    puts("st-info --flash");
    puts("st-info --sram");
    puts("st-info --descr");
    puts("st-info --pagesize");
    puts("st-info --chipid");
    puts("st-info --serial");
    puts("st-info --hla-serial");
    puts("st-info --probe");
}

/* Print normal or OpenOCD hla_serial with newline */
static void stlink_print_serial(stlink_t *sl, bool openocd)
{
    const char *fmt;

    if (openocd) {
       printf("\"");
       fmt = "\\x%02x";
    } else {
       fmt = "%02x";
    }

    for (int n = 0; n < sl->serial_size; n++)
        printf(fmt, sl->serial[n]);

    if (openocd)
       printf("\"");
    printf("\n");
}

static void stlink_print_info(stlink_t *sl)
{
    const chip_params_t *params = NULL;

    if (!sl)
        return;

    printf(" serial: ");
    stlink_print_serial(sl, false);
    printf("openocd: ");
    stlink_print_serial(sl, true);

    printf("  flash: %zu (pagesize: %zu)\n", sl->flash_size, sl->flash_pgsz);
    printf("   sram: %zu\n",       sl->sram_size);
    printf(" chipid: 0x%.4x\n",    sl->chip_id);

    for (size_t i = 0; i < sizeof(devices) / sizeof(devices[0]); i++) {
        if (devices[i].chip_id == sl->chip_id) {
            params = &devices[i];
            break;
        }
    }

    if (params)
        printf("  descr: %s\n", params->description);
}

static void stlink_probe(void)
{
    stlink_t **stdevs;
    size_t size;

    size = stlink_probe_usb(&stdevs);

    printf("Found %zu stlink programmers\n", size);

    for (size_t n = 0; n < size; n++)
        stlink_print_info(stdevs[n]);

    stlink_probe_usb_free(&stdevs, size);
}

static stlink_t *stlink_open_first(void)
{
    stlink_t* sl = NULL;
    sl = stlink_v1_open(0, 1);
    if (sl == NULL)
        sl = stlink_open_usb(0, 1, NULL);

    return sl;
}

static int print_data(char **av)
{
    stlink_t* sl = NULL;

    // Probe needs all devices unclaimed
    if (strcmp(av[1], "--probe") == 0) {
        stlink_probe();
        return 0;
    }

    sl = stlink_open_first();

    if (sl == NULL) {
        return -1;
    }

    sl->verbose = 0;

    if (stlink_current_mode(sl) == STLINK_DEV_DFU_MODE)
        stlink_exit_dfu_mode(sl);

    if (stlink_current_mode(sl) != STLINK_DEV_DEBUG_MODE)
        stlink_enter_swd_mode(sl);

    if (strcmp(av[1], "--flash") == 0)
        printf("0x%zx\n", sl->flash_size);
    else if (strcmp(av[1], "--sram") == 0)
        printf("0x%zx\n", sl->sram_size);
    else if (strcmp(av[1], "--pagesize") == 0)
        printf("0x%zx\n", sl->flash_pgsz);
    else if (strcmp(av[1], "--chipid") == 0)
        printf("0x%.4x\n", sl->chip_id);
    else if (strcmp(av[1], "--serial") == 0)
        stlink_print_serial(sl, false);
    else if (strcmp(av[1], "--hla-serial") == 0)
        stlink_print_serial(sl, true);
    else if (strcmp(av[1], "--descr") == 0) {
        const chip_params_t *params = NULL;
        for (size_t i = 0; i < sizeof(devices) / sizeof(devices[0]); i++) {
            if(devices[i].chip_id == sl->chip_id) {
                params = &devices[i];
                break;
            }
        }
        if (params == NULL) {
            return -1;
        }
        printf("%s\n", params->description);
    }

    if (sl)
    {
        stlink_exit_debug_mode(sl);
        stlink_close(sl);
    }

    return 0;
}

int main(int ac, char** av)
{
    int err = -1;
    if (ac < 2) {
        usage();
        return -1;
    }

    err = print_data(av);

    return err;
}
